Remove stripLeadingAndTrailingHTMLSpaces()

annevk opened the pull request:
#### ea527fa9c22d43cf4075f0364a6327a569a14a54
<pre>
Remove stripLeadingAndTrailingHTMLSpaces()
<a href="https://bugs.webkit.org/show_bug.cgi?id=257414">https://bugs.webkit.org/show_bug.cgi?id=257414</a>
rdar://109923008

Reviewed by Chris Dumez and Darin Adler.

Replace it with trim(isASCIIWhitespace),
isAllSpecialCharacters&lt;isASCIIWhitespace&lt;UChar&gt;&gt;, or remove it
altogether when the result is passed to WTF::protocolIsJavaScript() or
completeURL().

Also sprinkle a few FIXMEs around as we can probably do even better in
the future.

* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::invalidStatus const):
(WebCore::AccessibilityObject::currentState const):
(WebCore::AccessibilityObject::roleDescription const):
* Source/WebCore/dom/DataTransfer.cpp:
(WebCore::normalizeType):
(WebCore::DataTransfer::getDataForItem const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::processBaseElement):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeContainsJavaScriptURL const):
(WebCore::Element::getNonEmptyURLAttribute const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestClassicScript):
(WebCore::ScriptElement::requestModuleScript):
(WebCore::ScriptElement::requestImportMap):
(WebCore::ScriptElement::isScriptForEventSupported const):
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::create):
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::EmailInputType::typeMismatchFor const):
(WebCore::EmailInputType::sanitizeValue const):
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::attributeChanged):
(WebCore::HTMLAnchorElement::handleClick):
* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLEmbedElement.cpp:
(WebCore::HTMLEmbedElement::attributeChanged):
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::attributeChanged):
(WebCore::HTMLFrameElementBase::setLocation):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::extractMIMETypeFromTypeAttributeForLookup):
* Source/WebCore/html/HTMLImageLoader.cpp:
(WebCore::HTMLImageLoader::sourceURI const):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::attributeChanged):
(WebCore::HTMLObjectElement::parametersForPlugin):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::text const):
(WebCore::HTMLOptionElement::value const):
(WebCore::HTMLOptionElement::label const):
(WebCore::HTMLOptionElement::displayLabel const):
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLTablePartElement.cpp:
(WebCore::HTMLTablePartElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::imageSourceURL const):
(WebCore::HTMLVideoElement::posterImageURL const):
* Source/WebCore/html/URLInputType.cpp:
(WebCore::URLInputType::sanitizeValue const):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::stripLeadingAndTrailingHTMLSpaces): Deleted.
* Source/WebCore/html/parser/HTMLParserIdioms.h:
* Source/WebCore/loader/FormSubmission.cpp:
(WebCore::FormSubmission::Attributes::parseAction):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::extractContentLanguageFromHeader):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::decode):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::allowScriptWithNonce const):
(WebCore::ContentSecurityPolicy::allowStyleWithNonce const):
(WebCore::ContentSecurityPolicy::allowNonParserInsertedScripts const):
(WebCore::ContentSecurityPolicy::allowInlineScript const):
(WebCore::ContentSecurityPolicy::allowInlineStyle const):
(WebCore::ContentSecurityPolicy::allowScriptFromSource const):
(WebCore::ContentSecurityPolicy::allowStyleFromSource const):
* Source/WebCore/platform/ContentType.cpp:
(WebCore::ContentType::containerType const):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::defaultEventHandler):
* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::attributeContainsJavaScriptURL const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::performActionOnElement):

Canonical link: <a href="https://commits.webkit.org/264713@main">https://commits.webkit.org/264713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d4509ceb701e9b2b6c5bbc9ae1570999d701c33

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8356 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8649 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8866 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10024 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8421 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8365 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10637 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8560 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11285 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8502 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9561 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7575 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10205 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6859 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7646 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15232 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7980 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7789 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11174 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6735 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7550 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2033 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11760 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8004 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/dom/DataTransfer.cpp:
```diff
@@ -128,7 +127,7 @@ static String normalizeType(const String& type)
         return type;
 
-    String lowercaseType = stripLeadingAndTrailingHTMLSpaces(type).convertToASCIILowercase();
+    String lowercaseType = type.trim(isASCIIWhitespace).convertToASCIILowercase();
```
```suggestion
    auto lowercaseType = type.trim(isASCIIWhitespace).convertToASCIILowercase();
```

cdumez commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -315,7 +314,7 @@ bool ScriptElement::requestClassicScript(const String& sourceURL)
     ASSERT(m_element.isConnected());
     ASSERT(!m_loadableScript);
-    if (!stripLeadingAndTrailingHTMLSpaces(sourceURL).isEmpty()) {
+    if (!StringView(sourceURL).trim(isASCIIWhitespace<UChar>).isEmpty()) {
```
I think I'd use `!StringView(sourceURL).isAllSpecialCharacters<isASCIIWhitespace<UChar>>()` for such checks.

cdumez commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -354,7 +353,7 @@ bool ScriptElement::requestModuleScript(const TextPosition& scriptStartPosition)
 
         String sourceURL = sourceAttributeValue();
-        if (stripLeadingAndTrailingHTMLSpaces(sourceURL).isEmpty()) {
+        if (StringView(sourceURL).trim(isASCIIWhitespace<UChar>).isEmpty()) {
```
ditto

cdumez commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -399,7 +398,7 @@ bool ScriptElement::requestImportMap(LocalFrame& frame, const String& sourceURL)
     ASSERT(m_element.isConnected());
     ASSERT(!m_loadableScript);
-    if (!stripLeadingAndTrailingHTMLSpaces(sourceURL).isEmpty()) {
+    if (!StringView(sourceURL).trim(isASCIIWhitespace<UChar>).isEmpty()) {
```
ditto

cdumez commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -241,8 +240,8 @@ bool HTMLVideoElement::isURLAttribute(const Attribute& attribute) const
-    const AtomString& url = attributeWithoutSynchronization(posterAttr);
-    if (!stripLeadingAndTrailingHTMLSpaces(url).isEmpty())
+    const auto& url = attributeWithoutSynchronization(posterAttr);
+    if (!StringView(url).trim(isASCIIWhitespace<UChar>).isEmpty())
```
Could rely on `isAllSpecialCharacters()`

cdumez commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -184,7 +183,7 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
     // an empty string.
     CachedResourceHandle<CachedImage> newImage = nullptr;
-    if (!attr.isNull() && !stripLeadingAndTrailingHTMLSpaces(attr).isEmpty()) {
+    if (!attr.isNull() && !StringView(attr).trim(isASCIIWhitespace<UChar>).isEmpty()) {
```
Could use isAllSpecialCharacters

cdumez commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -488,8 +487,8 @@ void ImageLoader::decode(Ref<DeferredPromise>&& promise)
-    AtomString attr = element().imageSourceURL();
-    if (stripLeadingAndTrailingHTMLSpaces(attr).isEmpty()) {
+    auto attr = element().imageSourceURL();
+    if (StringView(attr).trim(isASCIIWhitespace<UChar>).isEmpty()) {
```
ditto

cdumez reviewed (approved):
LGTM with suggestions

darinadler commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -613,11 +612,11 @@ bool ScriptElement::isScriptForEventSupported() const
     String forAttribute = forAttributeValue();
     if (!eventAttribute.isNull() && !forAttribute.isNull()) {
-        forAttribute = stripLeadingAndTrailingHTMLSpaces(forAttribute);
+        forAttribute = forAttribute.trim(isASCIIWhitespace);
```
This should really be `StringView` since we only use it to check equality.

darinadler commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -613,11 +612,11 @@ bool ScriptElement::isScriptForEventSupported() const
             return false;
 
-        eventAttribute = stripLeadingAndTrailingHTMLSpaces(eventAttribute);
+        eventAttribute = eventAttribute.trim(isASCIIWhitespace);
```
Same here.

darinadler commented on Source/WebCore/html/EmailInputType.cpp:
```diff
@@ -68,7 +68,7 @@ bool EmailInputType::typeMismatchFor(const String& value) const
         return !isValidEmailAddress(value);
     for (auto& address : value.splitAllowingEmptyEntries(',')) {
-        if (!isValidEmailAddress(stripLeadingAndTrailingHTMLSpaces(address)))
+        if (!isValidEmailAddress(address.trim(isASCIIWhitespace)))
```
Would be better if this used `StringView`. Seems the string is not going to be stored.

darinadler reviewed (approved):
If we can use a `StringView` and never convert back into a `String` in even more places then it’s great to take advantage of the no-allocation `StringView::trim` function.

it’s even better to use `atomString.string()` than `String(atomString)`.

annevk committed:
Remove stripLeadingAndTrailingHTMLSpaces()
https://bugs.webkit.org/show_bug.cgi?id=257414
rdar://109923008

Reviewed by Chris Dumez and Darin Adler.

Replace it with trim(isASCIIWhitespace),
isAllSpecialCharacters<isASCIIWhitespace<UChar>>, or remove it
altogether when the result is passed to WTF::protocolIsJavaScript() or
completeURL().

Also sprinkle a few FIXMEs around as we can probably do even better in
the future.

* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::invalidStatus const):
(WebCore::AccessibilityObject::currentState const):
(WebCore::AccessibilityObject::roleDescription const):
* Source/WebCore/dom/DataTransfer.cpp:
(WebCore::normalizeType):
(WebCore::DataTransfer::getDataForItem const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::processBaseElement):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeContainsJavaScriptURL const):
(WebCore::Element::getNonEmptyURLAttribute const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestClassicScript):
(WebCore::ScriptElement::requestModuleScript):
(WebCore::ScriptElement::requestImportMap):
(WebCore::ScriptElement::isScriptForEventSupported const):
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::create):
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::EmailInputType::typeMismatchFor const):
(WebCore::EmailInputType::sanitizeValue const):
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::attributeChanged):
(WebCore::HTMLAnchorElement::handleClick):
* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLEmbedElement.cpp:
(WebCore::HTMLEmbedElement::attributeChanged):
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::attributeChanged):
(WebCore::HTMLFrameElementBase::setLocation):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::extractMIMETypeFromTypeAttributeForLookup):
* Source/WebCore/html/HTMLImageLoader.cpp:
(WebCore::HTMLImageLoader::sourceURI const):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::attributeChanged):
(WebCore::HTMLObjectElement::parametersForPlugin):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::text const):
(WebCore::HTMLOptionElement::value const):
(WebCore::HTMLOptionElement::label const):
(WebCore::HTMLOptionElement::displayLabel const):
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLTablePartElement.cpp:
(WebCore::HTMLTablePartElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::imageSourceURL const):
(WebCore::HTMLVideoElement::posterImageURL const):
* Source/WebCore/html/URLInputType.cpp:
(WebCore::URLInputType::sanitizeValue const):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::stripLeadingAndTrailingHTMLSpaces): Deleted.
* Source/WebCore/html/parser/HTMLParserIdioms.h:
* Source/WebCore/loader/FormSubmission.cpp:
(WebCore::FormSubmission::Attributes::parseAction):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::extractContentLanguageFromHeader):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::decode):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::allowScriptWithNonce const):
(WebCore::ContentSecurityPolicy::allowStyleWithNonce const):
(WebCore::ContentSecurityPolicy::allowNonParserInsertedScripts const):
(WebCore::ContentSecurityPolicy::allowInlineScript const):
(WebCore::ContentSecurityPolicy::allowInlineStyle const):
(WebCore::ContentSecurityPolicy::allowScriptFromSource const):
(WebCore::ContentSecurityPolicy::allowStyleFromSource const):
* Source/WebCore/platform/ContentType.cpp:
(WebCore::ContentType::containerType const):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::defaultEventHandler):
* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::attributeContainsJavaScriptURL const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::performActionOnElement):

Canonical link: https://commits.webkit.org/264713@main
